File: C++/Gun/Gun.cpp
#include <iostream>

#include "Gun.h"

using namespace std;

Gun::Gun(const string& model, int capacity) {
    this->amount = 0;
    this->capacity = capacity;
    this->isReady = false;
    this->model = model;
    this->totalShots = 0;
}

Gun::~Gun(){}

int Gun::getAmount() const {
    return this->amount;
}

int Gun::getCapacity() const {
    return this->capacity;
}

bool Gun::ready() const {
    return !isReady;
}

const string& Gun::getModel() const {
    return this->model;
}

int Gun::getTotalShots() const {
    return this->totalShots;
}

void Gun::prepare() {
    isReady = !isReady;
}

void Gun::reload() {
    amount = capacity;
}

void Gun::shoot() {
    if ( amount == 0 ) {
        throw OutOfRoundsException();
    }
    amount -= 1;
    totalShots += 1;
    cout << "Bang!" << endl;
}

ostream& operator<<(ostream& out, const Gun& gun) {
    out << "Gun: " << gun.getModel() << endl;
    out << "capacity: " << gun.getCapacity() << endl;
    out << "preparedness: ";
    if ( gun.ready() ) {
        out << "ready" << endl;
    } else {
        out << "not ready" << endl;
    }
    out << "rounds: (" << gun.getAmount();
    out << "/" << gun.getCapacity() << ")" << endl;
    out << "total: " << gun.getTotalShots() << endl;
    
    return out;
}

File: Java/src/com/gymfox/bcwc5/gun/Gun.java
package com.gymfox.bcwc5.gun;

public class Gun {
    private int amount;
    private int capacity;
    private boolean isReady;
    private String model;
    private int totalShots;

    public Gun(String model, int capacity) {
        this.amount = 0;
        this.capacity = capacity;
        this.isReady = false;
        this.model = model;
        this.totalShots = 0;
    }

    public void shoot() throws OutOfRoundsException, GunIsNotReadyException {
        if ( amount == 0 ) {
            throw new OutOfRoundsException();
        }
        if ( !ready() ) {
            throw new GunIsNotReadyException();
        }

        amount -= 1;
        totalShots += 1;

        System.out.println("Bang!");
        String.format("Total shots: ( %d / %d )", getTotalShots(), getCapacity());
    }

    public String toString() {
        StringBuffer out = new StringBuffer();

        out.append("Gun - " + getModel() +"\n")
                .append("Capacity / amount: ( " + getCapacity() + " / " + getAmount() + " )\n")
                .append(!ready() ? ("not ready\n") : ("ready\n"));

        return out.toString();
    }

    public Gun clone() throws CloneNotSupportedException {
        return (Gun) super.clone();
    }

    public void reload() {
        amount = capacity;
    }

    public void prepare() {
        isReady = !isReady;
    }

    public boolean ready() {
        return isReady;
    }

    public int getAmount() {
        return amount;
    }

    public int getCapacity() {
        return capacity;
    }

    public String getModel() {
        return model;
    }

    public int getTotalShots() {
        return totalShots;
    }
}